explorer/interpreter/type_checker.cpp

```diff
@@ -649,6 +780,22 @@ auto TypeChecker::ArgumentDeduction(
                                           << "actual: " << *arg;
     }
     const Value* subst_param_type = Substitute(deduced, param);
+    if (param->kind() == Value::Kind::VariableType) {
+      const auto& binding = cast<VariableType>(*param).binding();
+      if (binding.has_static_type()) {
+        if (binding.static_type().kind() != Value::Kind::TypeType &&
+            binding.static_type().kind() != Value::Kind::ConstraintType &&
+            binding.static_type().kind() != Value::Kind::InterfaceType) {
```

Comment:
A comment explaining why we exclude these cases would be useful. I'm not sure if this is "TODO: We don't implement proper convertibility checking for types yet" or if it's "don't check constraint types because they'll be checked separately by the caller later" or something else.

Can you use `IsTypeOfType` here?